### PR TITLE
fix: ensure transactions are never reported as both pending and confirmed

### PR DESCRIPTION
**Pull request by zone117x**
Fixes https://github.com/hirosystems/stacks-blockchain-api/issues/1557, related to https://github.com/hirosystems/stacks-blockchain-api/issues/1428

This adds a function which detects any transactions that are erroneously marked as both pending in the mempool table and also confirmed in the mined txs table.

The function is called after all other sql writes during block & microblock ingestion, but before any reads or view refreshes that depend on the mempool table.

This should be a safe work-around for whatever bug is causing the underlying problem.

### CODE REVIEW

File: src/datastore/pg-write-store.ts [hunk 1]
```diff
@@ -727,6 +731,10 @@ export class PgWriteStore extends PgStore {
         );
       }
 
+      if (!this.isEventReplay) {
+        await this.reconcileMempoolStatus(sql);
```
Comment (zone117x) [inline src/datastore/pg-write-store.ts:735]: @rafaelcr do these 2 call sites look safe to you?

Comment (rafaelcr) [reply src/datastore/pg-write-store.ts:735]: Yep, it's inside the write tx so I think we're good.

File: src/datastore/pg-write-store.ts [hunk 2]
```diff
@@ -450,6 +450,10 @@ export class PgWriteStore extends PgStore {
         }
       }
 
+      if (!this.isEventReplay) {
+        await this.reconcileMempoolStatus(sql);
+      }
+
       if (!this.isEventReplay) {
```
Comment (rafaelcr) [inline src/datastore/pg-write-store.ts:457]: Could you consolidate these two code blocks into the same conditional? Same in the other usage

File: src/datastore/pg-write-store.ts [hunk 3]
```diff
@@ -759,6 +767,34 @@ export class PgWriteStore extends PgStore {
     }
   }
 
+  // Find any transactions that are erroneously still marked as both `pending` in the mempool table
+  // and also confirmed in the mined txs table. Mark these as pruned in the mempool and log warning.
+  // This must be called _after_ any writes to txs/mempool tables during block and microblock ingestion,
+  // but _before_ any reads or view refreshes that depend on the mempool table.
+  // NOTE: this is essentially a work-around for whatever bug is causing the underlying problem.
+  async reconcileMempoolStatus(sql: PgSqlClient): Promise<void> {
```
Comment (rafaelcr) [inline src/datastore/pg-write-store.ts:775]: While this works great, I think we should add another similar check inside `updateMempoolTxs` that inserts them as `pruned` initially if we already have that same incoming mempool tx in a block (in a scenario where we receive a mempool tx with some huge delay for any reason).

That way, we won't display the inconsistent result from the time that delayed mempool tx arrives until the next block/microblock comes and we reconcile here.

What do you think?

File: src/tests/mempool-tests.ts [hunk 4]
```diff
@@ -1405,4 +1405,125 @@ describe('mempool tests', () => {
...
+      execution_cost_read_length: 0,
+      execution_cost_runtime: 0,
+      execution_cost_write_count: 0,
+      execution_cost_write_length: 0,
+    };
+
+    // Simulate the bug with a txs being in the mempool at confirmed at the same time by
+    // directly inserting the mempool-tx and mined-tx, bypassing the normal update functions.
+    await db.updateBlock(db.sql, dbBlock1);
+    await db.updateMempoolTxs({ mempoolTxs: [mempoolTx] });
```
Comment (rafaelcr) [inline src/tests/mempool-tests.ts:1490]: If we apply the fix to `updateMempoolTxs` too, you could still run this test just by inserting the tx directly into the DB here. Or maybe reversing the order of these two lines.

Comment (zone117x) [reply src/tests/mempool-tests.ts]: Good idea, done!

**Review by rafaelcr — APPROVED** [review]
LGTM!

**Pull request merged**